samples/Samples.Jwt/JwtWebSocketAuthenticationService.cs

```diff
@@ -0,0 +1,66 @@
+using System.Security.Claims;
+using GraphQL.Server.Transports.AspNetCore.WebSockets;
+using GraphQL.Transport;
+using Microsoft.AspNetCore.Authentication.JwtBearer;
+
+namespace GraphQL.Server.Samples.Jwt;
+
+/// <summary>
+/// Authenticates WebSocket connections via the 'payload' of the initialization packet.
+/// This is necessary because WebSocket connections initiated from the browser cannot
+/// authenticate via HTTP headers.
+/// <br/><br/>
+/// This class is not used when authenticating over GET/POST.
+/// </summary>
+public class JwtWebSocketAuthenticationService : IWebSocketAuthenticationService
+{
+    private readonly IGraphQLSerializer _graphQLSerializer;
+
+    public JwtWebSocketAuthenticationService(IGraphQLSerializer graphQLSerializer)
+    {
+        _graphQLSerializer = graphQLSerializer;
+    }
+
+    public Task AuthenticateAsync(IWebSocketConnection connection, string subProtocol, OperationMessage operationMessage)
+    {
+        try
+        {
+            // for connections authenticated via HTTP headers, no need to reauthenticate
+            if (connection.HttpContext.User.Identity?.IsAuthenticated ?? false)
+                return Task.CompletedTask;
+
+            // attempt to read the 'Authorization' key from the payload object and verify it contains "Bearer: XXXXXXXX"
+            var authPayload = _graphQLSerializer.ReadNode<AuthPayload>(operationMessage.Payload);
+            if (authPayload != null && authPayload.Authorization != null && authPayload.Authorization.StartsWith("Bearer ", StringComparison.Ordinal))
+            {
+                // pull the token from the value
+                var token = authPayload.Authorization.Substring(7);
+                // parse the token in the same manner that the .NET AddJwtBearer() method does
+                var handler = new Microsoft.IdentityModel.JsonWebTokens.JsonWebTokenHandler();
+                var result = handler.ValidateToken(token, JwtHelper.Instance.TokenValidationParameters);
+                if (result.IsValid)
+                {
+                    // convert JWT tokens with "role" as the claim type (a JWT defined claim type) to the proper .NET claim type constant
```

Comment:
What is _proper .NET claim type constant_ here?

Reply:
See: https://docs.microsoft.com/en-us/dotnet/api/system.security.claims.claimsidentity.roleclaimtype?view=net-6.0

The constant is: `http://schemas.microsoft.com/ws/2008/06/identity/claims/role`.

See: https://docs.microsoft.com/en-us/dotnet/api/system.security.claims.claimtypes.role?view=net-6.0

The ASP.NET JWT code does this conversion, I believe, but since we must recreate similar code for use with WebSockets, I've done so where described by the comments there.  If this conversion does not take place, .NET will not recognize the JWT roles as roles when evaluating via `IsInRole`.

Please keep in mind this is a _sample_ implementation.

Reply:
Understood, thanks.